### PR TITLE
Add id generating orm bucket

### PR DESCRIPTION
**Pull request by alpe**
Extract common id generating functionality into a new bucket type.

### CODE REVIEW

**Comment by ruseinov** [comment]
This looks good! I felt the need for something like this when implementing new extensions with Sequencers as there's a lot of boilerplate currently involved and we need to define a bunch of constants too.

File: orm/idgen_bucket.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import (
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+)
+
+// IDGenerator defines an interface for custom id generators
+type IDGenerator interface {
+	// NextVal returns a new unique ID key
+	NextVal(db weave.KVStore, obj CloneableData) ([]byte, error)
```
Comment (ethanfrey) [inline orm/idgen_bucket.go]: Why do you take obj here? If the nextval is just based on the bucket?

File: x/distribution/model.go [hunk 2]
```diff
@@ -102,30 +102,17 @@ func copyAddr(a weave.Address) weave.Address {
...
+	orm.IDGenBucket
 }
 
 // NewRevenueBucket returns a bucket for managing revenues state.
 func NewRevenueBucket() *RevenueBucket {
 	b := migration.NewBucket("distribution", "revenue", orm.NewSimpleObj(nil, &Revenue{}))
 	return &RevenueBucket{
-		Bucket: b,
-		idSeq:  b.Sequence("id"),
+		IDGenBucket: orm.WithSeqIDGenerator(b, "id"),
```
Comment (ethanfrey) [inline x/distribution/model.go]: Nice simplification here

File: orm/idgen_bucket.go (hunk 1 above)
Comment (alpe) [reply orm/idgen_bucket.go]: I had the idea of a generic IDGenerator interface that supports natural keys as well.

**Pull request merged**